## 문제 접근 방법

1. 반복문의 경우는 Bottom-up(작은 문제에서 출발)
2. 재귀의 경우는 Top-down (큰 문제에서 출발)

* 재귀의 경우에는 꼭 종료조건이 있어야 한다. (무한반복될 수 있어서)

In [2]:
# 0부터 10까지 올라가면서 더한 것 
x = 0
for i in range(11) :
    x += i
print(x)

55


In [4]:
# 10부터 0까지 내려가면서 더한 것 
def f(n):
    if n <=1:
        return 1
    else:
        return n + f(n-1)
print(f(10))

55


In [5]:
#무한루프 - 반복문
while True:
    if input('##') == 'exit':
        break
    if input('##') == 'hi':
        print('hello world')
    else : 
        continue

##hi
##hi
hello world
##hi
##hi
hello world
##exit


In [6]:
#무한루프 - 재귀함수
def console():
    if input('##') == 'exit':
        return None
    if input('##') == 'hi':
        print('hello world')
    console()
console()

#무한루프도 구현할 수 있다. But 꼭!!! 종료조건을 넣어야 한다.

##s
##d
##f
##e
##hi
##hi
hello world
##hi
##hi
hello world
##hi
##hi
hello world
##exit


In [12]:
#2진수 구하기 - 반복문
# print(bin(11)[2:])
# print(oct(11))
# print(hex(11))


x = int(input('2진수로 바꿀 숫자를 입력하세요:'))
result = ''

while True:
    if x % 2 == 0:
        result += '0'
    else:
        result += '1'
    x = x // 2
    if x == 1 or x == 0 :
        result += str(x)
        print(result[::-1])
        break

2진수로 바꿀 숫자를 입력하세요:11
1011


1. 2     11     1
2. 2      5     1
3. 2      2     0
4.        1

In [14]:
def 이진수구하기(입력값):
    if 입력값 < 2 :
        return str(입력값)
    else:
        return str(이진수구하기(입력값//2)) + str(입력값 % 2)
        
이진수구하기(11)

'1011'

이진수구하기(11) -> str(이진수구하기(5)) + str(1)   -> 1011 <br>
이진수구하기(5)  -> str(이진수구하기(2)) + str(1)   -> 101 <br>
이진수구하기(2)  -> str(이진수구하기(1)) + str(0)   -> 10 <br>
이진수구하기(1)  -> 1 

In [17]:
def 문자열뒤집기(문자열):
    if 문자열 == '' :
        return None
    else :
        문자열뒤집기(문자열[1:])
        print(문자열[0])
        
print(문자열뒤집기('leesuji'))

i
j
u
s
e
e
l
None


In [21]:
result = ''
result1 = ''
for i in 'leesuji':
    result+= i
    result1 = i + result1
print(result)
print(result1)

leesuji
ijuseel


In [23]:
x=0
for i in '2231':
    x += int(i)
print(x)

8


In [28]:
def 숫자더하기(문자열):
    if 문자열 == '' :
        return 0
    else:
        return int(문자열[0])+int(숫자더하기(문자열[1:]))
    
print(숫자더하기('2231'))

8
